crates/subspace-service/src/dsn/import_blocks/root_blocks.rs

```diff
@@ -68,121 +71,168 @@ impl RootBlockHandler {
+                                Ok(RootBlockResponse { root_blocks }) => {
+                                    trace!(
+                                        %peer_id,
+                                        root_blocks_number=%root_blocks.len(),
+                                        "Last root block request succeeded."
+                                    );
+
+                                    peer_blocks.borrow_mut().insert(peer_id, root_blocks);
+                                }
+                                Err(error) => {
+                                    debug!(%peer_id, ?error, "Last root block request failed.");
+                                }
+                            };
                         };
-                    };
 
-                    peer_block_tasks.push(peer_block_fut);
+                        peer_block_tasks.push(Box::pin(peer_block_fut));
+                    }
                 }
-            }
-            Err(err) => {
-                warn!(?err, "get_closest_peers returned an error");
+                Err(err) => {
+                    warn!(?err, "get_closest_peers returned an error");
 
-                return Err(err.into());
+                    return Err(err.into());
+                }
             }
-        }
 
-        join_all(peer_block_tasks).await;
+            join_all(peer_block_tasks).await;
 
-        let peer_blocks = Arc::try_unwrap(peer_blocks)
-            .expect("We manually waited for each other usage to be dropped.")
-            .into_inner();
-
-        // TODO: Consider adding attempts to increase the initial peer set.
-        if peer_blocks.len() < ROOT_BLOCK_CONSENSUS_MIN_SET {
-            return Err(format!(
-                "Root block consensus failed: not enough peers ({}).",
-                peer_blocks.len()
-            )
-            .into());
-        }
+            let peer_blocks = Arc::try_unwrap(peer_blocks)
+                .expect("We manually waited for each other usage to be dropped.")
+                .into_inner();
 
-        // Calculate votes
-        let mut root_block_score: HashMap<RootBlock, u64> = HashMap::new();
+            // TODO: Consider adding attempts to increase the initial peer set.
+            if peer_blocks.len() < ROOT_BLOCK_CONSENSUS_MIN_SET {
+                debug!(
+                    "Root block consensus failed: not enough peers ({}).",
+                    peer_blocks.len()
+                );
 
-        for (peer_id, root_blocks) in &peer_blocks {
-            if !self.validate_root_blocks(*peer_id, root_blocks).await {
                 continue;
             }
 
-            for root_block in root_blocks {
-                root_block_score
-                    .entry(*root_block)
-                    .and_modify(|val| *val += 1)
-                    .or_insert(1);
+            // Calculate votes
+            let mut root_block_score: HashMap<RootBlock, (u64, BTreeSet<PeerId>)> = HashMap::new();
```

Comment:
Nitpick, but this should be just a vector and not `BTreeSet`, we are not supposed to get multiple of the same block from the same peer unless business logic is very buggy.